src/box/box.cc

```diff
@@ -2632,54 +2649,70 @@ box_collect_confirmed_vclock(struct vclock *confirmed_vclock, double deadline)
 	if (vclock_count <= 1)
 		return 0;
 
-	struct sync_trigger_data data = {
-		.vclock_syncs = {0},
-		.collected_vclock_map = 0,
-		.waiter = fiber(),
-		.vclock = confirmed_vclock,
-		.deadline = deadline,
-		.count = vclock_count,
-		.is_timed_out = false,
-	};
-	bit_set(&data.collected_vclock_map, instance_id);
+	struct sync_trigger_data *data = (sync_trigger_data *)
+		mempool_alloc(&sync_trigger_data_pool);
+	memset(data->vclock_syncs, 0, sizeof(data->vclock_syncs));
+	data->collected_vclock_map = 0;
+	data->waiter = fiber();
+	data->vclock = confirmed_vclock;
+	data->deadline = deadline;
+	data->count = vclock_count;
+	data->is_timed_out = false;
+	/*
+	 * Let relay_trigger_vclock_sync know that the data is needed in
+	 * box_collect_confirmed_vclock. Otherwise, relay_trigger_vclock_sync
+	 * may free the data while this fiber yielding in
+	 * box_collect_confirmed_vclock.
+	 */
+	data->ref_count = 1;
+
+	bit_set(&data->collected_vclock_map, instance_id);
 	struct trigger on_relay_thread_start;
-	trigger_create(&on_relay_thread_start, relay_get_sync_on_start, &data,
+	trigger_create(&on_relay_thread_start, relay_get_sync_on_start, data,
 		       NULL);
 	trigger_add(&replicaset.on_relay_thread_start, &on_relay_thread_start);
 	struct trigger on_ack;
-	trigger_create(&on_ack, check_vclock_sync_on_ack, &data, NULL);
+	trigger_create(&on_ack, check_vclock_sync_on_ack, data, NULL);
 	trigger_add(&replicaset.on_ack, &on_ack);
+
+	auto guard = make_scoped_guard([&] {
+		trigger_clear(&on_ack);
+		trigger_clear(&on_relay_thread_start);
+		/*
+		 * box_collect_confirmed_vclock no longer uses data.
+		 * If no one else is using the data, free it.
+		 */
```

Comment:
Nit: I'd not use prefix decrement here, so we can add asertion, that `ref_count` after decrement is >= 0. Just to be sure)

```c
--data->ref_count;
assert(data->ref_count >= 0);
if (data->ref_count == 0)
    ...
```

Same in other place